{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ca4d9d7-2cf5-44cf-b701-e875ec88ae6f",
   "metadata": {},
   "source": [
    "## Sorting and observing metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f2be2620-e15c-47bf-a015-bfa7058e356e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config Completer.use_jedi = False\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6392b77f-9d4c-44ad-a743-66322858bf42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ID', 'original_genome_id', 'old_species_names', 'species_names_fig2',\n",
      "       'hex', 'id_table S1', 'Species name', 'Major clade', 'clade_order',\n",
      "       'Family', 'Genus', 'CUG usage'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>original_genome_id</th>\n",
       "      <th>old_species_names</th>\n",
       "      <th>species_names_fig2</th>\n",
       "      <th>hex</th>\n",
       "      <th>id_table S1</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "      <th>clade_order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>CUG usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>saprochaete_clavata</td>\n",
       "      <td>Saprochaete_clavata</td>\n",
       "      <td>Saprochaete_clavata</td>\n",
       "      <td>#FF2800</td>\n",
       "      <td>22</td>\n",
       "      <td>Saprochaete clavata</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "      <td>3</td>\n",
       "      <td>Dipodascaceae</td>\n",
       "      <td>Saprochaete</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>yHMPu5000034974_deakozyma_indianensis_160519</td>\n",
       "      <td>yHMPu5000034974_Deakozyma_indianensis</td>\n",
       "      <td>Deakozyma_indianensis</td>\n",
       "      <td>#FF2800</td>\n",
       "      <td>39</td>\n",
       "      <td>Deakozyma indianensis</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "      <td>3</td>\n",
       "      <td>Saccharomycetales incertae sedis</td>\n",
       "      <td>Deakozyma</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>clavispora_lusitaniae</td>\n",
       "      <td>Clavispora_lusitaniae</td>\n",
       "      <td>Clavispora_lusitaniae</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>135</td>\n",
       "      <td>Clavispora lusitaniae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Metschnikowiaceae</td>\n",
       "      <td>Clavispora</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>yHMPu5000034999_cephaloascus_fragrans_160519</td>\n",
       "      <td>yHMPu5000034999_Cephaloascus_fragrans</td>\n",
       "      <td>Cephaloascus_fragrans</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>204</td>\n",
       "      <td>Cephaloascus fragrans</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Cephaloascaceae</td>\n",
       "      <td>Cephaloascus</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>yHMPu5000034998_cephaloascus_albidus_160519</td>\n",
       "      <td>yHMPu5000034998_Cephaloascus_albidus</td>\n",
       "      <td>Cephaloascus_albidus</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>205</td>\n",
       "      <td>Cephaloascus albidus</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Cephaloascaceae</td>\n",
       "      <td>Cephaloascus</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella complicata</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>340</td>\n",
       "      <td>sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>340</td>\n",
       "      <td>Sclerotinia sclerotiorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>341</td>\n",
       "      <td>stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>341</td>\n",
       "      <td>Stagonospora nodorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>342</td>\n",
       "      <td>xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>342</td>\n",
       "      <td>Xylona heveae</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>343</td>\n",
       "      <td>schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>343</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>343 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                            original_genome_id  \\\n",
       "0      1                           saprochaete_clavata   \n",
       "1      2  yHMPu5000034974_deakozyma_indianensis_160519   \n",
       "2      3                         clavispora_lusitaniae   \n",
       "3      4  yHMPu5000034999_cephaloascus_fragrans_160519   \n",
       "4      5   yHMPu5000034998_cephaloascus_albidus_160519   \n",
       "..   ...                                           ...   \n",
       "338  339                          Saitoella_complicata   \n",
       "339  340                      sclerotinia_sclerotiorum   \n",
       "340  341                          stagonospora_nodorum   \n",
       "341  342                                 xylona_heveae   \n",
       "342  343                     schizosaccharomyces_pombe   \n",
       "\n",
       "                         old_species_names         species_names_fig2  \\\n",
       "0                      Saprochaete_clavata        Saprochaete_clavata   \n",
       "1    yHMPu5000034974_Deakozyma_indianensis      Deakozyma_indianensis   \n",
       "2                    Clavispora_lusitaniae      Clavispora_lusitaniae   \n",
       "3    yHMPu5000034999_Cephaloascus_fragrans      Cephaloascus_fragrans   \n",
       "4     yHMPu5000034998_Cephaloascus_albidus       Cephaloascus_albidus   \n",
       "..                                     ...                        ...   \n",
       "338                   Saitoella_complicata       Saitoella_complicata   \n",
       "339               Sclerotinia_sclerotiorum   Sclerotinia_sclerotiorum   \n",
       "340                   Stagonospora_nodorum       Stagonospora_nodorum   \n",
       "341                          Xylona_heveae              Xylona_heveae   \n",
       "342              Schizosaccharomyces_pombe  Schizosaccharomyces_pombe   \n",
       "\n",
       "         hex  id_table S1               Species name  \\\n",
       "0    #FF2800           22        Saprochaete clavata   \n",
       "1    #FF2800           39      Deakozyma indianensis   \n",
       "2    #FFd200          135      Clavispora lusitaniae   \n",
       "3    #FFd200          204      Cephaloascus fragrans   \n",
       "4    #FFd200          205       Cephaloascus albidus   \n",
       "..       ...          ...                        ...   \n",
       "338  #050f07          339       Saitoella complicata   \n",
       "339  #050f07          340   Sclerotinia sclerotiorum   \n",
       "340  #050f07          341       Stagonospora nodorum   \n",
       "341  #050f07          342              Xylona heveae   \n",
       "342  #050f07          343  Schizosaccharomyces pombe   \n",
       "\n",
       "                         Major clade  clade_order  \\\n",
       "0    Dipodascaceae/Trichomonascaceae            3   \n",
       "1    Dipodascaceae/Trichomonascaceae            3   \n",
       "2                           CUG-Ser1            8   \n",
       "3                           CUG-Ser1            8   \n",
       "4                           CUG-Ser1            8   \n",
       "..                               ...          ...   \n",
       "338                         outgroup            0   \n",
       "339                         outgroup            0   \n",
       "340                         outgroup            0   \n",
       "341                         outgroup            0   \n",
       "342                         outgroup            0   \n",
       "\n",
       "                               Family         Genus CUG usage  \n",
       "0                       Dipodascaceae   Saprochaete       Leu  \n",
       "1    Saccharomycetales incertae sedis     Deakozyma       Leu  \n",
       "2                   Metschnikowiaceae    Clavispora       Ser  \n",
       "3                     Cephaloascaceae  Cephaloascus       Ser  \n",
       "4                     Cephaloascaceae  Cephaloascus       Ser  \n",
       "..                                ...           ...       ...  \n",
       "338                          outgroup      outgroup       Leu  \n",
       "339                          outgroup      outgroup       Leu  \n",
       "340                          outgroup      outgroup       Leu  \n",
       "341                          outgroup      outgroup       Leu  \n",
       "342                          outgroup      outgroup       Leu  \n",
       "\n",
       "[343 rows x 12 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df = pd.read_csv('./data/343taxa_species-name_clade-name_color-code.txt', sep='\\t')\n",
    "print(species_df.columns)\n",
    "species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7af992b-7c27-4a75-be9d-f7b6ede29687",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Saprochaete clavata</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Deakozyma indianensis</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Clavispora lusitaniae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Cephaloascus fragrans</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Cephaloascus albidus</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID           Species name                      Major clade\n",
       "0   1    Saprochaete clavata  Dipodascaceae/Trichomonascaceae\n",
       "1   2  Deakozyma indianensis  Dipodascaceae/Trichomonascaceae\n",
       "2   3  Clavispora lusitaniae                         CUG-Ser1\n",
       "3   4  Cephaloascus fragrans                         CUG-Ser1\n",
       "4   5   Cephaloascus albidus                         CUG-Ser1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filtered_df = species_df[['ID', 'Species name', 'Major clade']]\n",
    "display(filtered_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "92778446-a95f-42a2-b663-d186c34e63f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>204</td>\n",
       "      <td>Lipomyces arxii</td>\n",
       "      <td>Lipomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>241</td>\n",
       "      <td>Cyberlindnera saturnus</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID            Species name       Major clade\n",
       "203  204         Lipomyces arxii    Lipomycetaceae\n",
       "240  241  Cyberlindnera saturnus  Phaffomycetaceae"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df[filtered_df['Species name'].str.lower().isin(['cyberlindnera saturnus', 'lipomyces arxii'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a1964d27-234c-4807-8ae7-66625683904d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>221</td>\n",
       "      <td>Candida vartiovaarae</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>222</td>\n",
       "      <td>Candida orba</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>223</td>\n",
       "      <td>Candida stellimalicola</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>224</td>\n",
       "      <td>Cyberlindnera fabianii</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>225</td>\n",
       "      <td>Cyberlindnera jadinii</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>226</td>\n",
       "      <td>Cyberlindnera americana</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>227</td>\n",
       "      <td>Komagataella populi</td>\n",
       "      <td>Pichiaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>228</td>\n",
       "      <td>Komagataella pseudopastoris</td>\n",
       "      <td>Pichiaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>229</td>\n",
       "      <td>Phaffomyces thermotolerans</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>230</td>\n",
       "      <td>Phaffomyces antillensis</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>231</td>\n",
       "      <td>Phaffomyces opuntiae</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>232</td>\n",
       "      <td>Starmera quercuum</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>233</td>\n",
       "      <td>Starmera amethionina</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>234</td>\n",
       "      <td>Wickerhamomyces anomalus</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>235</td>\n",
       "      <td>Wickerhamomyces hampshirensis</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>236</td>\n",
       "      <td>Wickerhamomyces canadensis</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>237</td>\n",
       "      <td>Wickerhamomyces ciferrii</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>238</td>\n",
       "      <td>Candida ponderosae</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>239</td>\n",
       "      <td>Barnettozyma pratensis</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>240</td>\n",
       "      <td>Barnettozyma hawaiiensis</td>\n",
       "      <td>Phaffomycetaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                   Species name       Major clade\n",
       "220  221           Candida vartiovaarae  Phaffomycetaceae\n",
       "221  222                   Candida orba  Phaffomycetaceae\n",
       "222  223         Candida stellimalicola  Phaffomycetaceae\n",
       "223  224         Cyberlindnera fabianii  Phaffomycetaceae\n",
       "224  225          Cyberlindnera jadinii  Phaffomycetaceae\n",
       "225  226        Cyberlindnera americana  Phaffomycetaceae\n",
       "226  227            Komagataella populi        Pichiaceae\n",
       "227  228    Komagataella pseudopastoris        Pichiaceae\n",
       "228  229     Phaffomyces thermotolerans  Phaffomycetaceae\n",
       "229  230        Phaffomyces antillensis  Phaffomycetaceae\n",
       "230  231           Phaffomyces opuntiae  Phaffomycetaceae\n",
       "231  232              Starmera quercuum  Phaffomycetaceae\n",
       "232  233           Starmera amethionina  Phaffomycetaceae\n",
       "233  234       Wickerhamomyces anomalus  Phaffomycetaceae\n",
       "234  235  Wickerhamomyces hampshirensis  Phaffomycetaceae\n",
       "235  236     Wickerhamomyces canadensis  Phaffomycetaceae\n",
       "236  237       Wickerhamomyces ciferrii  Phaffomycetaceae\n",
       "237  238             Candida ponderosae  Phaffomycetaceae\n",
       "238  239         Barnettozyma pratensis  Phaffomycetaceae\n",
       "239  240       Barnettozyma hawaiiensis  Phaffomycetaceae"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This shows that the txt species file's rows are not grouped by clade\n",
    "filtered_df.iloc[220:240]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1a777d79-af92-4741-af38-7c93ab1a09f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ID, Species name, Major clade]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get rid of outgroup species\n",
    "filtered_df = filtered_df[filtered_df['Major clade'].isin([\n",
    "'Pichiaceae',\n",
    "'Dipodascaceae/Trichomonascaceae',\n",
    "'Trigonopsidaceae',\n",
    "'CUG-Ser1',\n",
    "'CUG-Ser2',\n",
    "'Saccharomycodaceae',\n",
    "'Lipomycetaceae',\n",
    "'Phaffomycetaceae',\n",
    "'Saccharomycetaceae',\n",
    "'Sporopachydermia clade',\n",
    "'Alloascoideaceae',\n",
    "'CUG-Ala'\n",
    "])]\n",
    "\n",
    "filtered_df[filtered_df['Major clade'] == 'outgroup']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7768b090-07cd-45c3-8678-4cb99a45c4fd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>250</td>\n",
       "      <td>Alloascoidea hylecoeti</td>\n",
       "      <td>Alloascoideaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>111</td>\n",
       "      <td>Nakazawaea holstii</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>110</td>\n",
       "      <td>Nakazawaea peltata</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>127</td>\n",
       "      <td>Pachysolen tannophilus</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>Peterozyma xylosa</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>34</td>\n",
       "      <td>Peterozyma toletana</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>68</td>\n",
       "      <td>Metschnikowia hawaiiensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>74</td>\n",
       "      <td>Metschnikowia arizonensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>73</td>\n",
       "      <td>Metschnikowia ipomoeae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>72</td>\n",
       "      <td>Spathaspora gorwiae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>Spathaspora hagerdaliae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>70</td>\n",
       "      <td>Spathaspora girioi</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>69</td>\n",
       "      <td>Candida sojae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>67</td>\n",
       "      <td>Metschnikowia similis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>66</td>\n",
       "      <td>Metschnikowia bowlesiae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>64</td>\n",
       "      <td>Metschnikowia mauinuiana</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>75</td>\n",
       "      <td>Metschnikowia matae var. matae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>63</td>\n",
       "      <td>Metschnikowia proteae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>62</td>\n",
       "      <td>Metschnikowia kamakouana</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>61</td>\n",
       "      <td>Metschnikowia drakensbergensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                    Species name       Major clade\n",
       "249  250          Alloascoidea hylecoeti  Alloascoideaceae\n",
       "110  111              Nakazawaea holstii           CUG-Ala\n",
       "109  110              Nakazawaea peltata           CUG-Ala\n",
       "126  127          Pachysolen tannophilus           CUG-Ala\n",
       "32    33               Peterozyma xylosa           CUG-Ala\n",
       "33    34             Peterozyma toletana           CUG-Ala\n",
       "67    68       Metschnikowia hawaiiensis          CUG-Ser1\n",
       "73    74       Metschnikowia arizonensis          CUG-Ser1\n",
       "72    73          Metschnikowia ipomoeae          CUG-Ser1\n",
       "71    72             Spathaspora gorwiae          CUG-Ser1\n",
       "70    71         Spathaspora hagerdaliae          CUG-Ser1\n",
       "69    70              Spathaspora girioi          CUG-Ser1\n",
       "68    69                   Candida sojae          CUG-Ser1\n",
       "66    67           Metschnikowia similis          CUG-Ser1\n",
       "65    66         Metschnikowia bowlesiae          CUG-Ser1\n",
       "63    64        Metschnikowia mauinuiana          CUG-Ser1\n",
       "74    75  Metschnikowia matae var. matae          CUG-Ser1\n",
       "62    63           Metschnikowia proteae          CUG-Ser1\n",
       "61    62        Metschnikowia kamakouana          CUG-Ser1\n",
       "60    61  Metschnikowia drakensbergensis          CUG-Ser1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group clades together\n",
    "species_df_sorted = filtered_df.sort_values('Major clade')\n",
    "species_df_sorted.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e21bd767-9dfd-4398-8471-03c78577df02",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "CHC_records = list(SeqIO.parse('./data/CHC1.fasta', 'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "77735bd5-776b-4c29-a012-c0d21c22b342",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species in original txt file DataFrame: 343\n",
      "Number of species in filtered and sorted DataFrame: 332\n",
      "Number of species in unorganized CHC FASTA file: 345\n"
     ]
    }
   ],
   "source": [
    "# There are more species sequences in the CHC1.fasta than species in the informational DataFrame\n",
    "# Update: These extra sequences turned out to be duplicates and fragmented sequences\n",
    "print(f'Number of species in original txt file DataFrame: {len(species_df)}')\n",
    "print(f'Number of species in filtered and sorted DataFrame: {len(species_df_sorted)}')\n",
    "print(f'Number of species in unorganized CHC FASTA file: {len(CHC_records)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cbdf7cf-8999-437b-bfca-f260b9bac585",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Don't run these 2 cells again when rerunning the notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a6c6222f-429c-4014-b745-413e31b4b909",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHC1.fasta had multiple duplicates, so I created a file consisting only of them and aligned them\n",
    "from Bio import SeqIO\n",
    "import subprocess \n",
    "from io import StringIO \n",
    "from Bio.SeqRecord import SeqRecord \n",
    "\n",
    "duplicates = list(SeqIO.parse('./data/CHC1-duplicates.fasta', 'fasta'))\n",
    "output = StringIO()\n",
    "SeqIO.write(duplicates, output, 'fasta')\n",
    "duplicates_str = output.getvalue()\n",
    "\n",
    "child = subprocess.Popen(    \n",
    "    ['mafft', '--localpair', '--maxiterate', '1000', '--ep', '0', '--quiet', '-'], \n",
    "    stdin=subprocess.PIPE,    \n",
    "    stdout=subprocess.PIPE,   \n",
    "    stderr=subprocess.PIPE      \n",
    ")\n",
    "\n",
    "child_out, child_err = child.communicate(input=duplicates_str.encode())\n",
    "\n",
    "aligned_duplicates = list(SeqIO.parse(StringIO(child_out.decode()), \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ba6c70d7-89c3-49f9-b16c-7d8f4c6fe7fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./data/aligned-CHC1-duplicates.fasta', 'w') as output_handle:\n",
    "    SeqIO.write(aligned_duplicates, output_handle, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "484f0af5-5b7c-41c7-858e-c0f2bd7bfe03",
   "metadata": {},
   "source": [
    "### Grouping the cleaned CHC FASTA file's sequences by clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ed3fc9a-47bd-4eb9-9be4-f77374673055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "343\n",
      "332\n",
      "332\n"
     ]
    }
   ],
   "source": [
    "cleaned_CHC_records = list(SeqIO.parse('./data/CHC1-cleaned.fasta', 'fasta'))\n",
    "\n",
    "print(len(species_df))\n",
    "\n",
    "# The cleaned heavy chain dataset now has as many species' sequences as the informational dataset has species:\n",
    "print(len(species_df_sorted))\n",
    "print(len(cleaned_CHC_records))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "26aaf001-3393-4f49-ac0f-0a03e7e08c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>250</td>\n",
       "      <td>Alloascoidea hylecoeti</td>\n",
       "      <td>Alloascoideaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>111</td>\n",
       "      <td>Nakazawaea holstii</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110</td>\n",
       "      <td>Nakazawaea peltata</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>127</td>\n",
       "      <td>Pachysolen tannophilus</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33</td>\n",
       "      <td>Peterozyma xylosa</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID            Species name       Major clade\n",
       "0  250  Alloascoidea hylecoeti  Alloascoideaceae\n",
       "1  111      Nakazawaea holstii           CUG-Ala\n",
       "2  110      Nakazawaea peltata           CUG-Ala\n",
       "3  127  Pachysolen tannophilus           CUG-Ala\n",
       "4   33       Peterozyma xylosa           CUG-Ala"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df_sorted.reset_index(drop=True, inplace=True)\n",
    "species_df_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3cf9ec77-2186-4343-b5af-65d262b7578f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'yHMPu5000034754_lipomyces_arxii_160519_Seq_4699'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_CHC_records[1].id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "65bf4821-5c50-4f28-a23b-a721507dec6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species matched with CHC FASTA: 332\n",
      "Number of species not found in FASTA (species to be searched on BLAST or whose names are outdated): 0\n"
     ]
    }
   ],
   "source": [
    "species_order = species_df_sorted['Species name'].tolist()\n",
    "\n",
    "formatted_species_order = []\n",
    "for species_name in species_order:\n",
    "    species_name = species_name.lower().replace(' ', '_')\n",
    "    formatted_species_order.append(species_name)\n",
    "\n",
    "clade_order = species_df_sorted['Major clade'].tolist()\n",
    "ordered_CHCs = {}\n",
    "unmatched_species = []\n",
    "\n",
    "for species_name, clade in zip(formatted_species_order, clade_order):\n",
    "    matched = False\n",
    "    for record in cleaned_CHC_records:\n",
    "        if species_name in record.id:\n",
    "            ordered_CHCs[species_name] = record, clade\n",
    "            matched = True\n",
    "            break\n",
    "    if not matched:\n",
    "        print('Species not matched:', species_name)\n",
    "        unmatched_species.append(species_name)\n",
    "\n",
    "print('Number of species matched with CHC FASTA:', len(ordered_CHCs))\n",
    "print('Number of species not found in FASTA (species to be searched on BLAST or whose names are outdated):', len(unmatched_species))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d290cb47-046a-40ea-ab65-c14b6c8630e2",
   "metadata": {},
   "source": [
    "## Fixing outdated naming issues in CHC fasta file\n",
    "\n",
    "This section of cells may show confusing outputs or have somewhat redundant blocks of code. This is because these cells intended to solve an issue with outdated names and finding one sequence whose name seems to be completely absent in the '343taxa' txt dataset in the CHC fasta file, and the code merely provided me with crucial information at the time that I was fixing the issues. There's no point in creating a completely new fasta file for each time I update the species names; thus, the current outputs are not the outputs I originally had when fixing the problem. These current outputs may seem unimportant or confusing, but they actually show that the issue has been fixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6a02af95-3798-4cb2-bdd3-ce6bedff7f0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "332\n",
      "332\n",
      "332\n"
     ]
    }
   ],
   "source": [
    "print(len(formatted_species_order))\n",
    "print(len(cleaned_CHC_records))\n",
    "print(len(ordered_CHCs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4feea960-f820-494f-a96f-d722e6ce8f26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>original_genome_id</th>\n",
       "      <th>old_species_names</th>\n",
       "      <th>species_names_fig2</th>\n",
       "      <th>hex</th>\n",
       "      <th>id_table S1</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "      <th>clade_order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>CUG usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>333</td>\n",
       "      <td>arthrobotrys_oligospora</td>\n",
       "      <td>Arthrobotrys_oligospora</td>\n",
       "      <td>Arthrobotrys_oligospora</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>333</td>\n",
       "      <td>Arthrobotrys oligospora</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>334</td>\n",
       "      <td>aspergillus_nidulans</td>\n",
       "      <td>Aspergillus_nidulans</td>\n",
       "      <td>Aspergillus_nidulans</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>334</td>\n",
       "      <td>Aspergillus nidulans</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>335</td>\n",
       "      <td>botrytis_cinerea</td>\n",
       "      <td>Botrytis_cinerea</td>\n",
       "      <td>Botrytis_cinerea</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>335</td>\n",
       "      <td>Botrytis cinerea</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>336</td>\n",
       "      <td>Coccidioides_immitis</td>\n",
       "      <td>Coccidioides_immitis</td>\n",
       "      <td>Coccidioides_immitis</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>336</td>\n",
       "      <td>Coccidioides immitis</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>337</td>\n",
       "      <td>fusarium_graminearum</td>\n",
       "      <td>Fusarium_graminearum</td>\n",
       "      <td>Fusarium_graminearum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>337</td>\n",
       "      <td>Fusarium graminearum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>338</td>\n",
       "      <td>neurospora_crassa</td>\n",
       "      <td>Neurospora_crassa</td>\n",
       "      <td>Neurospora_crassa</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>338</td>\n",
       "      <td>Neurospora crassa</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella complicata</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>340</td>\n",
       "      <td>sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>340</td>\n",
       "      <td>Sclerotinia sclerotiorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>341</td>\n",
       "      <td>stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>341</td>\n",
       "      <td>Stagonospora nodorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>342</td>\n",
       "      <td>xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>342</td>\n",
       "      <td>Xylona heveae</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>343</td>\n",
       "      <td>schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>343</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID         original_genome_id          old_species_names  \\\n",
       "332  333    arthrobotrys_oligospora    Arthrobotrys_oligospora   \n",
       "333  334       aspergillus_nidulans       Aspergillus_nidulans   \n",
       "334  335           botrytis_cinerea           Botrytis_cinerea   \n",
       "335  336       Coccidioides_immitis       Coccidioides_immitis   \n",
       "336  337       fusarium_graminearum       Fusarium_graminearum   \n",
       "337  338          neurospora_crassa          Neurospora_crassa   \n",
       "338  339       Saitoella_complicata       Saitoella_complicata   \n",
       "339  340   sclerotinia_sclerotiorum   Sclerotinia_sclerotiorum   \n",
       "340  341       stagonospora_nodorum       Stagonospora_nodorum   \n",
       "341  342              xylona_heveae              Xylona_heveae   \n",
       "342  343  schizosaccharomyces_pombe  Schizosaccharomyces_pombe   \n",
       "\n",
       "            species_names_fig2      hex  id_table S1  \\\n",
       "332    Arthrobotrys_oligospora  #050f07          333   \n",
       "333       Aspergillus_nidulans  #050f07          334   \n",
       "334           Botrytis_cinerea  #050f07          335   \n",
       "335       Coccidioides_immitis  #050f07          336   \n",
       "336       Fusarium_graminearum  #050f07          337   \n",
       "337          Neurospora_crassa  #050f07          338   \n",
       "338       Saitoella_complicata  #050f07          339   \n",
       "339   Sclerotinia_sclerotiorum  #050f07          340   \n",
       "340       Stagonospora_nodorum  #050f07          341   \n",
       "341              Xylona_heveae  #050f07          342   \n",
       "342  Schizosaccharomyces_pombe  #050f07          343   \n",
       "\n",
       "                  Species name Major clade  clade_order    Family     Genus  \\\n",
       "332    Arthrobotrys oligospora    outgroup            0  outgroup  outgroup   \n",
       "333       Aspergillus nidulans    outgroup            0  outgroup  outgroup   \n",
       "334           Botrytis cinerea    outgroup            0  outgroup  outgroup   \n",
       "335       Coccidioides immitis    outgroup            0  outgroup  outgroup   \n",
       "336       Fusarium graminearum    outgroup            0  outgroup  outgroup   \n",
       "337          Neurospora crassa    outgroup            0  outgroup  outgroup   \n",
       "338       Saitoella complicata    outgroup            0  outgroup  outgroup   \n",
       "339   Sclerotinia sclerotiorum    outgroup            0  outgroup  outgroup   \n",
       "340       Stagonospora nodorum    outgroup            0  outgroup  outgroup   \n",
       "341              Xylona heveae    outgroup            0  outgroup  outgroup   \n",
       "342  Schizosaccharomyces pombe    outgroup            0  outgroup  outgroup   \n",
       "\n",
       "    CUG usage  \n",
       "332       Leu  \n",
       "333       Leu  \n",
       "334       Leu  \n",
       "335       Leu  \n",
       "336       Leu  \n",
       "337       Leu  \n",
       "338       Leu  \n",
       "339       Leu  \n",
       "340       Leu  \n",
       "341       Leu  \n",
       "342       Leu  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df[species_df['Major clade'] == 'outgroup']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "77b20c1b-0704-4165-8e31-88a458df42ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['arthrobotrys_oligospora', 'aspergillus_nidulans', 'botrytis_cinerea', 'coccidioides_immitis', 'fusarium_graminearum', 'neurospora_crassa', 'saitoella_complicata', 'sclerotinia_sclerotiorum', 'stagonospora_nodorum', 'xylona_heveae', 'schizosaccharomyces_pombe']\n",
      "Outgroup species not matched: arthrobotrys_oligospora\n",
      "Outgroup species not matched: aspergillus_nidulans\n",
      "Outgroup species not matched: botrytis_cinerea\n",
      "Outgroup species not matched: coccidioides_immitis\n",
      "Outgroup species not matched: fusarium_graminearum\n",
      "Outgroup species not matched: neurospora_crassa\n",
      "Outgroup species not matched: saitoella_complicata\n",
      "Outgroup species not matched: sclerotinia_sclerotiorum\n",
      "Outgroup species not matched: stagonospora_nodorum\n",
      "Outgroup species not matched: xylona_heveae\n",
      "Outgroup species not matched: schizosaccharomyces_pombe\n"
     ]
    }
   ],
   "source": [
    "outgroup_species = species_df[species_df['Major clade'] == 'outgroup']\n",
    "\n",
    "outgroup_species = outgroup_species['Species name'].tolist()\n",
    "\n",
    "formatted_outgroup_species = []\n",
    "for species_name in outgroup_species:\n",
    "    species_name = species_name.lower().replace(' ', '_')\n",
    "    formatted_outgroup_species.append(species_name)\n",
    "\n",
    "print(formatted_outgroup_species)\n",
    "\n",
    "for species in formatted_outgroup_species:\n",
    "    matched = False\n",
    "    for record in cleaned_CHC_records:\n",
    "        if species in record.id:\n",
    "            print('Outgroup species matched:', species)\n",
    "            matched = True\n",
    "            break\n",
    "    if not matched:\n",
    "        print('Outgroup species not matched:', species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a36ebd92-3cbc-4ed3-85ae-d942645d9a3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>sequence</th>\n",
       "      <th>id</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alloascoidea_hylecoeti</td>\n",
       "      <td>MSNSSAPPSRCGIWEVDDFSDCFRHSYIETYFPLIILSLSVLFILI...</td>\n",
       "      <td>alloascoidea_hylecoeti_Seq_6913</td>\n",
       "      <td>alloascoidea_hylecoeti_Seq_6913 y1000_id=0_691...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nakazawaea_holstii</td>\n",
       "      <td>MGSEIPIDFTELAQLTQLGIQPQSLDFRSTTLESDKYVCIREQTAN...</td>\n",
       "      <td>yHMPu5000034918_nakazawaea_holstii_160519_Seq_...</td>\n",
       "      <td>yHMPu5000034918_nakazawaea_holstii_160519_Seq_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nakazawaea_peltata</td>\n",
       "      <td>MSDIPIEFTELAQLTALGVAPLSLEFRSTTLESDRYVCVREQGAGG...</td>\n",
       "      <td>nakazawaea_peltata_Seq_3684</td>\n",
       "      <td>nakazawaea_peltata_Seq_3684 y1000_id=99_3683 g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pachysolen_tannophilus</td>\n",
       "      <td>MASEIPIDFTELAQLTSLGIQSQSLDFKSTTLESDKYVCVRESTAN...</td>\n",
       "      <td>pachysolen_tannophilus_Seq_5663</td>\n",
       "      <td>pachysolen_tannophilus_Seq_5663 y1000_id=106_5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>peterozyma_xylosa</td>\n",
       "      <td>MSNDIPIDFTELAQLTSLGIQPNSLDFMSTTLESDKYVCIRETTAS...</td>\n",
       "      <td>yHMPu5000034883_peterozyma_xylosa_160519_Seq_2679</td>\n",
       "      <td>yHMPu5000034883_peterozyma_xylosa_160519_Seq_2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  species                                           sequence  \\\n",
       "0  alloascoidea_hylecoeti  MSNSSAPPSRCGIWEVDDFSDCFRHSYIETYFPLIILSLSVLFILI...   \n",
       "1      nakazawaea_holstii  MGSEIPIDFTELAQLTQLGIQPQSLDFRSTTLESDKYVCIREQTAN...   \n",
       "2      nakazawaea_peltata  MSDIPIEFTELAQLTALGVAPLSLEFRSTTLESDRYVCVREQGAGG...   \n",
       "3  pachysolen_tannophilus  MASEIPIDFTELAQLTSLGIQSQSLDFKSTTLESDKYVCVRESTAN...   \n",
       "4       peterozyma_xylosa  MSNDIPIDFTELAQLTSLGIQPNSLDFMSTTLESDKYVCIRETTAS...   \n",
       "\n",
       "                                                  id  \\\n",
       "0                    alloascoidea_hylecoeti_Seq_6913   \n",
       "1  yHMPu5000034918_nakazawaea_holstii_160519_Seq_...   \n",
       "2                        nakazawaea_peltata_Seq_3684   \n",
       "3                    pachysolen_tannophilus_Seq_5663   \n",
       "4  yHMPu5000034883_peterozyma_xylosa_160519_Seq_2679   \n",
       "\n",
       "                                         description  \n",
       "0  alloascoidea_hylecoeti_Seq_6913 y1000_id=0_691...  \n",
       "1  yHMPu5000034918_nakazawaea_holstii_160519_Seq_...  \n",
       "2  nakazawaea_peltata_Seq_3684 y1000_id=99_3683 g...  \n",
       "3  pachysolen_tannophilus_Seq_5663 y1000_id=106_5...  \n",
       "4  yHMPu5000034883_peterozyma_xylosa_160519_Seq_2...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for species, (seq_record, clade) in ordered_CHCs.items():\n",
    "    rows.append({\n",
    "        'species': species,\n",
    "        'sequence': str(seq_record.seq),\n",
    "        'id': seq_record.id,\n",
    "        'description': seq_record.description\n",
    "    })\n",
    "\n",
    "CHC_dict_df = pd.DataFrame(rows)\n",
    "CHC_dict_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "95b218d2-b3df-40d8-b233-438458e13644",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_CHCs_list = []\n",
    "for key in ordered_CHCs:\n",
    "    ordered_CHCs_list.append(str(ordered_CHCs[key][0].seq))\n",
    "\n",
    "CHCs_fasta_list = []\n",
    "for seq_record in cleaned_CHC_records:\n",
    "    CHCs_fasta_list.append(str(seq_record.seq))\n",
    "\n",
    "for ordered_seq in ordered_CHCs_list:\n",
    "    matched = False\n",
    "    for seq in CHCs_fasta_list:\n",
    "        if ordered_seq == seq:\n",
    "            matched = True\n",
    "            break\n",
    "    if not matched:\n",
    "        print('Sequence not found in ordered_CHCs_list:', seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ccf7da1b-0f91-4142-9034-afa54d72802c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 missing sequences\n"
     ]
    }
   ],
   "source": [
    "missing_count = 0\n",
    "missing_seqs = []\n",
    "for seq in CHCs_fasta_list:\n",
    "    matched = False\n",
    "    for ordered_seq in ordered_CHCs_list:\n",
    "        if ordered_seq == seq:\n",
    "            matched = True\n",
    "            break\n",
    "    if not matched:\n",
    "        print('Not found:', seq[0:100] + '...')\n",
    "        missing_seqs.append(seq)\n",
    "        missing_count += 1\n",
    "print(f'{missing_count} missing sequences')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a71f3da5-08b7-44b5-9cca-9093e178b05c",
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_seq_names = []\n",
    "for missing_seq in missing_seqs:\n",
    "    for seq_record in cleaned_CHC_records:\n",
    "        if missing_seq == seq_record.seq:\n",
    "            print(seq_record.id)\n",
    "            missing_seq_names.append(seq_record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "57b88697-fc20-47ea-89d6-d37131cf39f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "\n",
      "Number names in fasta file that aren`t in the \"Species name\" column of the txt file (outdated names): 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>original_genome_id</th>\n",
       "      <th>old_species_names</th>\n",
       "      <th>species_names_fig2</th>\n",
       "      <th>hex</th>\n",
       "      <th>id_table S1</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "      <th>clade_order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>CUG usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ID, original_genome_id, old_species_names, species_names_fig2, hex, id_table S1, Species name, Major clade, clade_order, Family, Genus, CUG usage]\n",
       "Index: []"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Many of the species names in the fasta file have an outdated version of their name. \n",
    "# I'm using this to get rid of them (and also potentially find that one missing corresponding sequence).\n",
    "formatted_missing_seq_names = []\n",
    "for name in missing_seq_names:\n",
    "    parts = name.split('_')\n",
    "    for part in parts:\n",
    "        if 'Seq' in part:\n",
    "            del parts[-2:]\n",
    "        elif part.startswith('yHMP'):\n",
    "            del parts[3]\n",
    "    name = '_'.join(parts)\n",
    "    print(name)\n",
    "    formatted_missing_seq_names.append(name.lower())\n",
    "\n",
    "print(formatted_missing_seq_names)\n",
    "print()\n",
    "\n",
    "old_names_in_fasta = species_df[species_df['old_species_names'].str.lower().isin(name for name in formatted_missing_seq_names)]\n",
    "print('Number names in fasta file that aren`t in the \"Species name\" column of the txt file (outdated names):', len(old_names_in_fasta))\n",
    "old_names_in_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e3636f81-c8a4-4ae9-ad4c-551d1c578e1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSDIPIEFTELTDLTSLGINQQSLEFRSTTLESDHYVVVREQVNGANTVAIVDLHNNNQVTRKNMSADNAILHPKQFIISLRANGTTLQIFNLETKEKLKSYNLEEPVIFWKWLNDEQLGLITATSIFTWNVFDGQPTAGPTKLTARHANLNNAQIINFVANKNFDWFAVVGITQENGRIAGKIQLYSKQRNVSQPIEGHVAGFGGITLEGASSPTQLFVCGNRTATGGQLHIIEIDHDNNNPHFQKKSVEIFFPPDATNDFPISIQISNKYGVIYLLTKYGFIHLYDLESGSNLFVNRITATPVFIASSYDDNNGILAINKTGQVLAVEISKDKLIPYILNKLANIPLALALASRGGLPGAENLFLQQFDTLLSQGDYANAAKVAASSEQLRTPQTIQKLKNVHAAPGAISPILQYFSTLLDKGKLNQFETIELAKPVLQQDRKQLFEKWLKEDKLTSSEELGDIVKPFDLNLALAIYLRAGTHAKVISALAELGQFDKIIPYSEKVGYQPNFIVLISNLLRSNPDKASEFAISLLNSPATSGQIEVEKISDIFFSQNFIQQGTSFLLDALKDDSPNQGHLQTRLLEINLLHAPQVADAILGNDMFHHYDRPTIAQLSEKAGLFQRALENYSDIKDIKRVIVHTNAIPADWLVAYFGKLNVEQSLVALRELLDKNIAQNLQIVIQVATKFSDLIGSATLIKLFEEFKSFEGLYYYLASTVNLTDDKDVVFKYIQSAAKLGQFKEIERVVKDNNVYDPEKVKNFLKDANLPDQLPLIIVCDRYNYVHDLILYLYKHQFFKFIEVYVQQVNPSKTAQVVAALLDVDCDEKVIQNLLQSVLGQVPVAELTTEVEKRNRLKLLLPFLEATLNSGSQDQAVFNTLAKIYIDSNNNPEKFLKENDSYDTKEVGHYCEKRDPYLAYIAYEKGKNDEELIRITNENSMYKYQARYLLARSDPSLWNVVLSEENIHRRQLIDQVVGVAVPEAVDAEPISLTVKAFMDNGLTSELIELLEKIILEPSPFNENPSLQGLLLLTAIKVEPTKVSSLVDKLDHFDPIEIGNLATEAGLHEEAFEIYDKFKLHSKALNVLVEDILSLDRGEAYAEKIGEAELWSQLGAAQLNGLRIPEAIDSYIKANDPSNYEAVIDIADSAGKFEELVKYLLAARQQLKEPKIDGELIVAYAELGKLNEIENLLNGSNVADLNEVGDKLYERKDYKAAKLLYSSVSNYSKLASTLVFLDDYQSAVECARKASNIKVWKQVNDACLEHKEFRLAQICGLNLIVHAEELDELVSKYETEGYFNELISLFEAGLGLERAHMGMFTELAILYTKYNPSRTLEHLKLFWSRINIPKVIRATEAAHLWPELIFLYAHYDEWDNAALTMIEKSSSSFDHSSFKEIIVKVANLEIYYKAINFYVNEHPSLITDLLAVLTARIDIPRVVRIFQKSDNLPLIKSFLISVLEKNNSVVNAAYHDLLIEEEDYETLRNVVDAHDKYDSIDLGARLEKHDLIFFRQIAAHLYRKNKKWNKALSILKADQLWKDAIETAAISQSTDVAEDLLTYFVETGNNEAFIGLLYSGYDLIRYDKVLELEWLHDLGDYIKPYLITIERERSENIQNLIKDLKIREEAKAKEETEPILGGQRLLITNGAPIGYQPTGQGFGNGF\n",
      "      ID         Species name       Major clade\n",
      "156  188  Wickerhamomyces sp.  Phaffomycetaceae\n"
     ]
    }
   ],
   "source": [
    "# Originally filtered this one out; turns out it, too, had an outdated name. The 3 cells below show it's properly added and ordered now:\n",
    "print(ordered_CHCs['wickerhamomyces_sp.'][0].seq)\n",
    "print(species_df_sorted[species_df_sorted['Species name'] == 'Wickerhamomyces sp.'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "273102a9-b08e-41cc-9e5b-fd4b48a62a68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wickerhamomyces_sp. (SeqRecord(seq=Seq('MSDIPIEFTELTDLTSLGINQQSLEFRSTTLESDHYVVVREQVNGANTVAIVDL...NGF'), id='yHMPu5000035286_wickerhamomyces_sp._160928_Seq_128', name='yHMPu5000035286_wickerhamomyces_sp._160928_Seq_128', description='yHMPu5000035286_wickerhamomyces_sp._160928_Seq_128 y1000_id=283_127 gene_full=genemark-NODE_26_length_125553_cov_16.6396_ID_2725-processed-gene-0.16 CDS=1-4983', dbxrefs=[]), 'Phaffomycetaceae')\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for key, value in ordered_CHCs.items():\n",
    "    i += 1\n",
    "    if i == 157:\n",
    "        print(key, value)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3e94274c-b62a-463e-970d-088543aa82fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID                              188\n",
       "Species name    Wickerhamomyces sp.\n",
       "Major clade        Phaffomycetaceae\n",
       "Name: 156, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df_sorted.iloc[156]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc75767e-9d58-429b-a5d4-6d4ca7ecab7d",
   "metadata": {},
   "source": [
    "## Confirming the order of the dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2f5829c8-590d-45f2-9b1c-2249ec01ac35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "332 matches found\n"
     ]
    }
   ],
   "source": [
    "match_count = 0\n",
    "match_found = False\n",
    "for species in formatted_species_order:\n",
    "    if species in ordered_CHCs:\n",
    "        match_found = True\n",
    "        match_count += 1\n",
    "\n",
    "if not match_found:\n",
    "    print('No matches found')\n",
    "\n",
    "print(f'{match_count} matches found')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a00d007d-3c2b-467c-b79d-7c2cdb20c61f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Same elements\n",
      "Elements are in the same order\n"
     ]
    }
   ],
   "source": [
    "dict_keys = list(ordered_CHCs.keys())\n",
    "\n",
    "if set(formatted_species_order) == set(dict_keys):\n",
    "    print('Same elements')\n",
    "    if formatted_species_order == dict_keys:\n",
    "        print('Elements are in the same order')\n",
    "    else:\n",
    "        print('Order differs')\n",
    "else:\n",
    "    print('Elements differ between the list and the dictionary keys.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "47562e9a-9a74-41ea-b1e9-fb0fee7ce36e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(CHC_dict_df['species'].tolist() == formatted_species_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3960dff-99bd-4e7c-8216-67c464e93c19",
   "metadata": {},
   "source": [
    "## Convert the ordered dictionary into a FASTA file for MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a385219e-4c75-419b-8f42-6e4b9e9d7cc7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Prepend clade name to each SeqRecord object's description to more easily split the aligned file into multiple FASTAs\n",
    "for species, (seq_record, clade) in ordered_CHCs.items():\n",
    "    clade = clade.replace('/', '_').replace(' ', '-')\n",
    "    seq_record.description = f'{clade} | {seq_record.description}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3aac0dd6-3f9f-4d40-9c51-0de72653d721",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 332 sequences to ordered_CHCs.fasta\n"
     ]
    }
   ],
   "source": [
    "seq_records = [seq_record for (seq_record, clade) in ordered_CHCs.values()]\n",
    "\n",
    "with open('./data/ordered-CHCs.fasta', 'w') as handle:\n",
    "    SeqIO.write(seq_records, handle, 'fasta')\n",
    "\n",
    "# NOTE: ordered-CHCs.fasta has the human CHC as well\n",
    "print(f'Wrote {len(seq_records)} sequences to ordered_CHCs.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41c7da1d-aae4-4e3f-827d-2102e72f6935",
   "metadata": {},
   "source": [
    "### Note: Make sure to add the human CHC to ordered-CHCs.fasta before the following cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4f37d9b1-47e8-4926-9473-14a0a9f7115b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3287, 1672, 1674, 1667, 1668, 1668, 1665, 1658, 1665, 1523, 1673, 1673, 1673, 1665, 1665, 1665, 1665, 1665, 1665, 1665, 1665, 1665, 1665, 1665, 1666, 1665, 1665, 1665, 1665, 1667, 1657, 2016, 1666, 1666, 1667, 1666, 1666, 1668, 1676, 1667, 1661, 1662, 1671, 1664, 1658, 2015, 1662, 1665, 1665, 1658, 1665, 1660, 1665, 1665, 1669, 1669, 1669, 1661, 1667, 1664, 1664, 1664, 1662, 1665, 2018, 1662, 1635, 1673, 1669, 1635, 1660, 1671, 1663, 1668, 1666, 1606, 2055, 1669, 1663, 1661, 1665, 1658, 1667, 1664, 1660, 1668, 1668, 2033, 1667, 1672, 1667, 1668, 1676, 1667, 1667, 1667, 1667, 1672, 1668, 1662, 1706, 1691, 1719, 1692, 1643, 1647, 1643, 1670, 1643, 1647, 1670, 1670, 1643, 1643, 1675, 1660, 1674, 1678, 1670, 1656, 1660, 1655, 1670, 1676, 1653, 1672, 1669, 1643, 1656, 1662, 1642, 1648, 1663, 1656, 1643, 1427, 2199, 1680, 1654, 1668, 1666, 1659, 1660, 1655, 1660, 1662, 1661, 1660, 1671, 1658, 1662, 2034, 1662, 1660, 1661, 1662, 1661, 1661, 1663, 2085, 1666, 1663, 1661, 1660, 1661, 1661, 2076, 1662, 1662, 1662, 1666, 1662, 1662, 1663, 1661, 1661, 1667, 1661, 1663, 1664, 1661, 1615, 1667, 1664, 1653, 1655, 1654, 1656, 1670, 1664, 2076, 1823, 1670, 1663, 1656, 2005, 1667, 1666, 1666, 1662, 1662, 1662, 1663, 1663, 1662, 1635, 2258, 1662, 1662, 1660, 1662, 1660, 1663, 1662, 1667, 1665, 1666, 1668, 1665, 1666, 1666, 1664, 1666, 1666, 1649, 1666, 1665, 1667, 1666, 1309, 1993, 1665, 2024, 1662, 2315, 2011, 2317, 1661, 1669, 1660, 1659, 1666, 1658, 2318, 1670, 1648, 1648, 1664, 1653, 1652, 1652, 1648, 1655, 1619, 1652, 1647, 1652, 1653, 1653, 1653, 1653, 1652, 1651, 1658, 1654, 1657, 1655, 1653, 1626, 1644, 1655, 1652, 1652, 1652, 1652, 1650, 1655, 1653, 1655, 1654, 1663, 1654, 1648, 1654, 1653, 1654, 1652, 1649, 1649, 1651, 1651, 1654, 1653, 1647, 1655, 1653, 1655, 1654, 1650, 1657, 1654, 1657, 1590, 1652, 1652, 1644, 1657, 1657, 1657, 1659, 1658, 1659, 1657, 1658, 1658, 1654, 1641, 1658, 1643, 1643, 1643, 1643, 1658, 1643, 1670, 1671, 1673, 1675, 1668, 1671, 1671, 1656]\n"
     ]
    }
   ],
   "source": [
    "# These are the various CHC lengths. They vary quite a lot, so --localpair will be used with MAFFT.\n",
    "print([len(seq_record.seq) for seq_record in seq_records])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4a588c32-3070-4803-8f6e-ed61ed27d92a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment completed successfully\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "with open('./data/ordered-CHCs.fasta', 'r') as handle:\n",
    "    ordered_seqs_str = handle.read()\n",
    "\n",
    "cmd = [\n",
    "    'mafft',\n",
    "    '--localpair',\n",
    "    '--maxiterate', '1000',\n",
    "    '--thread', '10',\n",
    "    '--quiet',\n",
    "    '-'\n",
    "]\n",
    "\n",
    "process = subprocess.run(\n",
    "    cmd,\n",
    "    input=ordered_seqs_str,\n",
    "    text=True,\n",
    "    capture_output=True\n",
    ")\n",
    "\n",
    "if process.returncode != 0:\n",
    "    print('MAFFT error:', process.stderr)\n",
    "else:\n",
    "    aligned_fasta = process.stdout\n",
    "    print('Alignment completed successfully')\n",
    "    with open('./data/CHC1-total-aligned.fasta', 'w') as out_f:\n",
    "        out_f.write(aligned_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7664d7d7-1e7e-40c5-ad2b-0b7c789837ba",
   "metadata": {},
   "source": [
    "## Split aligned FASTA into FASTAs by clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f849c997-ed58-4052-b6fc-ee2eb477b4df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First take out the human alignment\n",
    "total_aligned_records = list(SeqIO.parse('./data/CHC1-total-aligned.fasta', 'fasta'))\n",
    "\n",
    "total_aligned_records = total_aligned_records[1:]\n",
    "\n",
    "with open('./data/CHC1-total-aligned.fasta', 'w') as output_handle:\n",
    "    SeqIO.write(total_aligned_records, output_handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7c248048-3932-482b-a83a-cf7ff27fde4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split into 12 files, each one representing a clade\n",
    "from collections import defaultdict\n",
    "\n",
    "clade_groups = defaultdict(list)\n",
    "\n",
    "with open('./data/CHC1-total-aligned.fasta') as handle:\n",
    "    for seq_record in SeqIO.parse(handle, 'fasta'):\n",
    "        clade_name = seq_record.description.split(' ', 1)[1].split('|')[0].strip()\n",
    "        clade_groups[clade_name].append(seq_record)\n",
    "\n",
    "for clade, seqs in clade_groups.items():\n",
    "    clade = clade.replace('/', '_').replace(' ', '-')\n",
    "    with open(f'./data/CHC1-fastas-by-clade/{clade}_aligned_CHCs.fasta', 'w') as out_f:\n",
    "        SeqIO.write(seqs, out_f, 'fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
