{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "300ca4b4-798e-4146-9705-22e97158629a",
   "metadata": {},
   "source": [
    "### Sorting and observing metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fafb7305-c667-4316-8268-df4ebc72ee4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config Completer.use_jedi = False\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "892c1e42-ace6-4cec-8d0b-c1d9577d8ab1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ID', 'original_genome_id', 'old_species_names', 'species_names_fig2',\n",
      "       'hex', 'id_table S1', 'Species name', 'Major clade', 'clade_order',\n",
      "       'Family', 'Genus', 'CUG usage'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>original_genome_id</th>\n",
       "      <th>old_species_names</th>\n",
       "      <th>species_names_fig2</th>\n",
       "      <th>hex</th>\n",
       "      <th>id_table S1</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "      <th>clade_order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>CUG usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>saprochaete_clavata</td>\n",
       "      <td>Saprochaete_clavata</td>\n",
       "      <td>Saprochaete_clavata</td>\n",
       "      <td>#FF2800</td>\n",
       "      <td>22</td>\n",
       "      <td>Saprochaete clavata</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "      <td>3</td>\n",
       "      <td>Dipodascaceae</td>\n",
       "      <td>Saprochaete</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>yHMPu5000034974_deakozyma_indianensis_160519</td>\n",
       "      <td>yHMPu5000034974_Deakozyma_indianensis</td>\n",
       "      <td>Deakozyma_indianensis</td>\n",
       "      <td>#FF2800</td>\n",
       "      <td>39</td>\n",
       "      <td>Deakozyma indianensis</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "      <td>3</td>\n",
       "      <td>Saccharomycetales incertae sedis</td>\n",
       "      <td>Deakozyma</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>clavispora_lusitaniae</td>\n",
       "      <td>Clavispora_lusitaniae</td>\n",
       "      <td>Clavispora_lusitaniae</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>135</td>\n",
       "      <td>Clavispora lusitaniae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Metschnikowiaceae</td>\n",
       "      <td>Clavispora</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>yHMPu5000034999_cephaloascus_fragrans_160519</td>\n",
       "      <td>yHMPu5000034999_Cephaloascus_fragrans</td>\n",
       "      <td>Cephaloascus_fragrans</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>204</td>\n",
       "      <td>Cephaloascus fragrans</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Cephaloascaceae</td>\n",
       "      <td>Cephaloascus</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>yHMPu5000034998_cephaloascus_albidus_160519</td>\n",
       "      <td>yHMPu5000034998_Cephaloascus_albidus</td>\n",
       "      <td>Cephaloascus_albidus</td>\n",
       "      <td>#FFd200</td>\n",
       "      <td>205</td>\n",
       "      <td>Cephaloascus albidus</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "      <td>8</td>\n",
       "      <td>Cephaloascaceae</td>\n",
       "      <td>Cephaloascus</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>Saitoella_complicata</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>339</td>\n",
       "      <td>Saitoella complicata</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>340</td>\n",
       "      <td>sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>Sclerotinia_sclerotiorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>340</td>\n",
       "      <td>Sclerotinia sclerotiorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>341</td>\n",
       "      <td>stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>Stagonospora_nodorum</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>341</td>\n",
       "      <td>Stagonospora nodorum</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>342</td>\n",
       "      <td>xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>Xylona_heveae</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>342</td>\n",
       "      <td>Xylona heveae</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>343</td>\n",
       "      <td>schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>Schizosaccharomyces_pombe</td>\n",
       "      <td>#050f07</td>\n",
       "      <td>343</td>\n",
       "      <td>Schizosaccharomyces pombe</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>0</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>outgroup</td>\n",
       "      <td>Leu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>343 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                            original_genome_id  \\\n",
       "0      1                           saprochaete_clavata   \n",
       "1      2  yHMPu5000034974_deakozyma_indianensis_160519   \n",
       "2      3                         clavispora_lusitaniae   \n",
       "3      4  yHMPu5000034999_cephaloascus_fragrans_160519   \n",
       "4      5   yHMPu5000034998_cephaloascus_albidus_160519   \n",
       "..   ...                                           ...   \n",
       "338  339                          Saitoella_complicata   \n",
       "339  340                      sclerotinia_sclerotiorum   \n",
       "340  341                          stagonospora_nodorum   \n",
       "341  342                                 xylona_heveae   \n",
       "342  343                     schizosaccharomyces_pombe   \n",
       "\n",
       "                         old_species_names         species_names_fig2  \\\n",
       "0                      Saprochaete_clavata        Saprochaete_clavata   \n",
       "1    yHMPu5000034974_Deakozyma_indianensis      Deakozyma_indianensis   \n",
       "2                    Clavispora_lusitaniae      Clavispora_lusitaniae   \n",
       "3    yHMPu5000034999_Cephaloascus_fragrans      Cephaloascus_fragrans   \n",
       "4     yHMPu5000034998_Cephaloascus_albidus       Cephaloascus_albidus   \n",
       "..                                     ...                        ...   \n",
       "338                   Saitoella_complicata       Saitoella_complicata   \n",
       "339               Sclerotinia_sclerotiorum   Sclerotinia_sclerotiorum   \n",
       "340                   Stagonospora_nodorum       Stagonospora_nodorum   \n",
       "341                          Xylona_heveae              Xylona_heveae   \n",
       "342              Schizosaccharomyces_pombe  Schizosaccharomyces_pombe   \n",
       "\n",
       "         hex  id_table S1               Species name  \\\n",
       "0    #FF2800           22        Saprochaete clavata   \n",
       "1    #FF2800           39      Deakozyma indianensis   \n",
       "2    #FFd200          135      Clavispora lusitaniae   \n",
       "3    #FFd200          204      Cephaloascus fragrans   \n",
       "4    #FFd200          205       Cephaloascus albidus   \n",
       "..       ...          ...                        ...   \n",
       "338  #050f07          339       Saitoella complicata   \n",
       "339  #050f07          340   Sclerotinia sclerotiorum   \n",
       "340  #050f07          341       Stagonospora nodorum   \n",
       "341  #050f07          342              Xylona heveae   \n",
       "342  #050f07          343  Schizosaccharomyces pombe   \n",
       "\n",
       "                         Major clade  clade_order  \\\n",
       "0    Dipodascaceae/Trichomonascaceae            3   \n",
       "1    Dipodascaceae/Trichomonascaceae            3   \n",
       "2                           CUG-Ser1            8   \n",
       "3                           CUG-Ser1            8   \n",
       "4                           CUG-Ser1            8   \n",
       "..                               ...          ...   \n",
       "338                         outgroup            0   \n",
       "339                         outgroup            0   \n",
       "340                         outgroup            0   \n",
       "341                         outgroup            0   \n",
       "342                         outgroup            0   \n",
       "\n",
       "                               Family         Genus CUG usage  \n",
       "0                       Dipodascaceae   Saprochaete       Leu  \n",
       "1    Saccharomycetales incertae sedis     Deakozyma       Leu  \n",
       "2                   Metschnikowiaceae    Clavispora       Ser  \n",
       "3                     Cephaloascaceae  Cephaloascus       Ser  \n",
       "4                     Cephaloascaceae  Cephaloascus       Ser  \n",
       "..                                ...           ...       ...  \n",
       "338                          outgroup      outgroup       Leu  \n",
       "339                          outgroup      outgroup       Leu  \n",
       "340                          outgroup      outgroup       Leu  \n",
       "341                          outgroup      outgroup       Leu  \n",
       "342                          outgroup      outgroup       Leu  \n",
       "\n",
       "[343 rows x 12 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df = pd.read_csv('./data/343taxa_species-name_clade-name_color-code.txt', sep='\\t')\n",
    "print(species_df.columns)\n",
    "species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9cdc590c-1002-4bb0-a978-4cfae7ec6709",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Saprochaete clavata</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Deakozyma indianensis</td>\n",
       "      <td>Dipodascaceae/Trichomonascaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Clavispora lusitaniae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Cephaloascus fragrans</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Cephaloascus albidus</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID           Species name                      Major clade\n",
       "0   1    Saprochaete clavata  Dipodascaceae/Trichomonascaceae\n",
       "1   2  Deakozyma indianensis  Dipodascaceae/Trichomonascaceae\n",
       "2   3  Clavispora lusitaniae                         CUG-Ser1\n",
       "3   4  Cephaloascus fragrans                         CUG-Ser1\n",
       "4   5   Cephaloascus albidus                         CUG-Ser1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filtered_df = species_df[['ID', 'Species name', 'Major clade']]\n",
    "display(filtered_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4ca77615-b46f-48d3-9c9d-c949b03d7b22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ID, Species name, Major clade]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df = filtered_df[filtered_df['Major clade'].isin([\n",
    "'Pichiaceae',\n",
    "'Dipodascaceae/Trichomonascaceae',\n",
    "'Trigonopsidaceae',\n",
    "'CUG-Ser1',\n",
    "'CUG-Ser2',\n",
    "'Saccharomycodaceae',\n",
    "'Lipomycetaceae',\n",
    "'Phaffomycetaceae',\n",
    "'Saccharomycetaceae',\n",
    "'Sporopachydermia clade',\n",
    "'Alloascoideaceae',\n",
    "'CUG-Ala'\n",
    "])]\n",
    "\n",
    "filtered_df[filtered_df['Major clade'] == 'outgroup']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea19a08b-9e3a-45fe-9cd5-6c55b943b766",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>250</td>\n",
       "      <td>Alloascoidea hylecoeti</td>\n",
       "      <td>Alloascoideaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>111</td>\n",
       "      <td>Nakazawaea holstii</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>110</td>\n",
       "      <td>Nakazawaea peltata</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>127</td>\n",
       "      <td>Pachysolen tannophilus</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>Peterozyma xylosa</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>34</td>\n",
       "      <td>Peterozyma toletana</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>68</td>\n",
       "      <td>Metschnikowia hawaiiensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>74</td>\n",
       "      <td>Metschnikowia arizonensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>73</td>\n",
       "      <td>Metschnikowia ipomoeae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>72</td>\n",
       "      <td>Spathaspora gorwiae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>71</td>\n",
       "      <td>Spathaspora hagerdaliae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>70</td>\n",
       "      <td>Spathaspora girioi</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>69</td>\n",
       "      <td>Candida sojae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>67</td>\n",
       "      <td>Metschnikowia similis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>66</td>\n",
       "      <td>Metschnikowia bowlesiae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>64</td>\n",
       "      <td>Metschnikowia mauinuiana</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>75</td>\n",
       "      <td>Metschnikowia matae var. matae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>63</td>\n",
       "      <td>Metschnikowia proteae</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>62</td>\n",
       "      <td>Metschnikowia kamakouana</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>61</td>\n",
       "      <td>Metschnikowia drakensbergensis</td>\n",
       "      <td>CUG-Ser1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID                    Species name       Major clade\n",
       "249  250          Alloascoidea hylecoeti  Alloascoideaceae\n",
       "110  111              Nakazawaea holstii           CUG-Ala\n",
       "109  110              Nakazawaea peltata           CUG-Ala\n",
       "126  127          Pachysolen tannophilus           CUG-Ala\n",
       "32    33               Peterozyma xylosa           CUG-Ala\n",
       "33    34             Peterozyma toletana           CUG-Ala\n",
       "67    68       Metschnikowia hawaiiensis          CUG-Ser1\n",
       "73    74       Metschnikowia arizonensis          CUG-Ser1\n",
       "72    73          Metschnikowia ipomoeae          CUG-Ser1\n",
       "71    72             Spathaspora gorwiae          CUG-Ser1\n",
       "70    71         Spathaspora hagerdaliae          CUG-Ser1\n",
       "69    70              Spathaspora girioi          CUG-Ser1\n",
       "68    69                   Candida sojae          CUG-Ser1\n",
       "66    67           Metschnikowia similis          CUG-Ser1\n",
       "65    66         Metschnikowia bowlesiae          CUG-Ser1\n",
       "63    64        Metschnikowia mauinuiana          CUG-Ser1\n",
       "74    75  Metschnikowia matae var. matae          CUG-Ser1\n",
       "62    63           Metschnikowia proteae          CUG-Ser1\n",
       "61    62        Metschnikowia kamakouana          CUG-Ser1\n",
       "60    61  Metschnikowia drakensbergensis          CUG-Ser1"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group clades together\n",
    "species_df_sorted = filtered_df.sort_values('Major clade')\n",
    "species_df_sorted.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8c37e8d8-0b4b-4484-9bcb-15f51dd0d520",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "CLC_records = list(SeqIO.parse('./data/CLC1.fasta', 'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "55a74a57-24ad-4819-94cf-25fa03431d9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species in original txt file DataFrame: 343\n",
      "Number of species in filtered and sorted DataFrame: 332\n",
      "Number of species in CLC FASTA file: 311\n"
     ]
    }
   ],
   "source": [
    "# There are some species whose light chains couldn't be found\n",
    "print(f'Number of species in original txt file DataFrame: {len(species_df)}')\n",
    "print(f'Number of species in filtered and sorted DataFrame: {len(species_df_sorted)}')\n",
    "print(f'Number of species in CLC FASTA file: {len(CLC_records)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5d5dde4f-5604-48b0-a4a5-c83be6a9c440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Major clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>250</td>\n",
       "      <td>Alloascoidea hylecoeti</td>\n",
       "      <td>Alloascoideaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>111</td>\n",
       "      <td>Nakazawaea holstii</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>110</td>\n",
       "      <td>Nakazawaea peltata</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>127</td>\n",
       "      <td>Pachysolen tannophilus</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33</td>\n",
       "      <td>Peterozyma xylosa</td>\n",
       "      <td>CUG-Ala</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID            Species name       Major clade\n",
       "0  250  Alloascoidea hylecoeti  Alloascoideaceae\n",
       "1  111      Nakazawaea holstii           CUG-Ala\n",
       "2  110      Nakazawaea peltata           CUG-Ala\n",
       "3  127  Pachysolen tannophilus           CUG-Ala\n",
       "4   33       Peterozyma xylosa           CUG-Ala"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df_sorted.reset_index(drop=True, inplace=True)\n",
    "species_df_sorted.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0f5c0ba-cdab-4f0d-b773-d2a2cba5cca5",
   "metadata": {},
   "source": [
    "### Grouping the CLC FASTA by clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "347a707f-95d6-473a-85a0-ec4c5aed7c8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species not matched: candida_sojae\n",
      "Species not matched: saturnispora_mendoncae\n",
      "Species not matched: kazachstania_transvaalensis\n",
      "Species not matched: hanseniaspora_clermontiae\n",
      "Number of species matched with CLC FASTA: 328\n",
      "Number of species to be searched on BLAST, whose names are outdated or misspelled, or lack a light chain in their genome: 4\n"
     ]
    }
   ],
   "source": [
    "cleaned_CLC_records = list(SeqIO.parse('./data/CLC1-cleaned.fasta', 'fasta'))\n",
    "\n",
    "species_order = species_df_sorted['Species name'].tolist()\n",
    "\n",
    "formatted_species_order = []\n",
    "for species_name in species_order:\n",
    "    species_name = species_name.lower().replace(' ', '_')\n",
    "    formatted_species_order.append(species_name)\n",
    "\n",
    "clade_order = species_df_sorted['Major clade'].tolist()\n",
    "ordered_CLCs = {}\n",
    "unmatched_species = []\n",
    "\n",
    "for species_name, clade in zip(formatted_species_order, clade_order):\n",
    "    matched = False\n",
    "    for record in cleaned_CLC_records:\n",
    "        if species_name in record.id:\n",
    "            ordered_CLCs[species_name] = record, clade\n",
    "            matched = True\n",
    "            break\n",
    "    if not matched:\n",
    "        print('Species not matched:', species_name)\n",
    "        unmatched_species.append(species_name)\n",
    "\n",
    "print('Number of species matched with CLC FASTA:', len(ordered_CLCs))\n",
    "print('Number of species to be searched on BLAST, whose names are outdated or misspelled, or lack a light chain in their genome:', len(unmatched_species))\n",
    "# These last 4 species don't have sequenced light chains in their known genome, so I'm just going to ignore them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed97587-577b-4c6c-bfc1-9ca8097b33c4",
   "metadata": {},
   "source": [
    "## Confirming the order of the dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7b0a32d7-2333-4887-9a90-6f1db68207e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "328\n"
     ]
    }
   ],
   "source": [
    "formatted_species_order = [\n",
    "    species for species in formatted_species_order if species not in [\n",
    "        'candida_sojae', \n",
    "        'saturnispora_mendoncae', \n",
    "        'kazachstania_transvaalensis', \n",
    "        'hanseniaspora_clermontiae'\n",
    "    ]\n",
    "]\n",
    "\n",
    "print(len(formatted_species_order))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "733bcada-ccce-41f3-867c-650f17d7fa53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Same elements\n",
      "Elements are in the same order\n"
     ]
    }
   ],
   "source": [
    "dict_keys = list(ordered_CLCs.keys())\n",
    "\n",
    "if set(formatted_species_order) == set(dict_keys):\n",
    "    print('Same elements')\n",
    "    if formatted_species_order == dict_keys:\n",
    "        print('Elements are in the same order')\n",
    "    else:\n",
    "        print('Order differs')\n",
    "else:\n",
    "    print('Elements differ between the list and the dictionary keys.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9d7688f5-7edc-4898-becf-40f7b23311ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for species, (seq_record, clade) in ordered_CLCs.items():\n",
    "    rows.append({\n",
    "        'species': species,\n",
    "        'sequence': str(seq_record.seq),\n",
    "        'id': seq_record.id,\n",
    "        'description': seq_record.description\n",
    "    })\n",
    "\n",
    "CLC_dict_df = pd.DataFrame(rows)\n",
    "CLC_dict_df.head()\n",
    "\n",
    "(CLC_dict_df['species'].tolist() == formatted_species_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c78c38d-6f03-4c6b-9c33-f089341caeee",
   "metadata": {},
   "source": [
    "## Convert the ordered dictionary into a FASTA file for MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f6e569f1-d8d0-4d4c-a6f8-4f0057df002b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepend clade name to each SeqRecord object's description to more easily split the aligned file into multiple FASTAs\n",
    "for species, (seq_record, clade) in ordered_CLCs.items():\n",
    "    clade = clade.replace('/', '_').replace(' ', '-')\n",
    "    seq_record.description = f'{clade} | {seq_record.description}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f01811de-f037-4a04-9150-d515a04a05db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 328 sequences to ordered_CLCs.fasta\n"
     ]
    }
   ],
   "source": [
    "seq_records = [seq_record for (seq_record, clade) in ordered_CLCs.values()]\n",
    "\n",
    "with open('./data/ordered-CLCs.fasta', 'w') as handle:\n",
    "    SeqIO.write(seq_records, handle, 'fasta')\n",
    "\n",
    "# NOTE: ordered-CLCs.fasta has the human CLC A manually added as well\n",
    "print(f'Wrote {len(seq_records)} sequences to ordered_CLCs.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d65ce569-7fd2-4e5d-8613-ab71bb2dcf00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[218, 221, 222, 221, 227, 215, 216, 213, 213, 213, 200, 218, 217, 213, 213, 213, 213, 212, 213, 212, 213, 214, 211, 211, 218, 213, 213, 213, 213, 217, 203, 221, 219, 217, 213, 217, 217, 207, 193, 192, 212, 213, 214, 210, 189, 210, 204, 213, 213, 210, 213, 220, 210, 208, 203, 218, 222, 221, 212, 203, 214, 214, 203, 214, 221, 214, 222, 229, 220, 222, 225, 225, 207, 227, 228, 209, 204, 217, 220, 225, 218, 210, 207, 214, 213, 212, 195, 216, 204, 215, 212, 217, 233, 218, 216, 217, 217, 221, 220, 204, 260, 217, 253, 232, 217, 217, 217, 239, 217, 236, 227, 226, 232, 217, 228, 225, 236, 228, 230, 225, 225, 228, 225, 227, 207, 224, 230, 217, 216, 222, 209, 230, 222, 212, 215, 222, 221, 231, 211, 220, 210, 244, 220, 213, 220, 226, 221, 220, 213, 225, 199, 209, 201, 199, 208, 198, 199, 197, 190, 204, 191, 191, 201, 194, 200, 194, 197, 196, 197, 200, 191, 197, 197, 201, 198, 209, 201, 199, 200, 204, 200, 200, 202, 194, 210, 212, 210, 218, 233, 217, 220, 217, 223, 227, 218, 214, 229, 230, 216, 269, 242, 230, 220, 221, 222, 218, 218, 233, 228, 231, 226, 221, 221, 237, 225, 220, 214, 214, 218, 220, 239, 226, 218, 219, 238, 235, 223, 220, 217, 212, 235, 212, 230, 216, 249, 218, 227, 243, 224, 218, 222, 217, 228, 233, 125, 125, 190, 188, 173, 166, 125, 167, 179, 180, 176, 236, 234, 236, 236, 234, 175, 172, 215, 215, 220, 234, 229, 120, 216, 178, 177, 170, 164, 200, 217, 204, 220, 212, 201, 212, 125, 233, 205, 215, 210, 157, 157, 146, 155, 229, 149, 234, 229, 201, 205, 229, 212, 216, 241, 216, 215, 210, 212, 211, 218, 214, 216, 217, 211, 218, 217, 217, 217, 235, 226, 226, 229, 217, 215, 229, 230, 235, 231, 223, 214, 213, 206, 205, 214]\n"
     ]
    }
   ],
   "source": [
    "# These are the various CLC lengths:\n",
    "# Note: I didn't just get the lengths of each object in the previously defined seq_records list because I had to tweak one of the yeast sequences later on.\n",
    "CLC_lengths = [len(record) for record in SeqIO.parse('./data/ordered-CLCs.fasta', 'fasta')]\n",
    "print(CLC_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "8955e5b6-a8bf-4674-ab2f-04801b2130a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment completed successfully\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "with open('./data/ordered-CLCs.fasta', 'r') as handle:\n",
    "    ordered_seqs_str = handle.read()\n",
    "\n",
    "cmd = [\n",
    "    'mafft',\n",
    "    '--localpair',\n",
    "    '--maxiterate', '1000',\n",
    "    '--thread', '10',\n",
    "    '--quiet',\n",
    "    '-'\n",
    "]\n",
    "\n",
    "process = subprocess.run(\n",
    "    cmd,\n",
    "    input=ordered_seqs_str,\n",
    "    text=True,\n",
    "    capture_output=True\n",
    ")\n",
    "\n",
    "if process.returncode != 0:\n",
    "    print('MAFFT error:', process.stderr)\n",
    "else:\n",
    "    aligned_fasta = process.stdout\n",
    "    print('Alignment completed successfully')\n",
    "    with open('./data/CLC1-total-aligned.fasta', 'w') as out_f:\n",
    "        out_f.write(aligned_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb1701b-2900-46be-bfed-a5f0121941fd",
   "metadata": {},
   "source": [
    "## Split aligned FASTA into FASTAs by clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "cdeb540d-2a31-488a-8d6e-11858928d069",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First take out the human alignment\n",
    "total_aligned_records = list(SeqIO.parse('./data/CLC1-total-aligned.fasta', 'fasta'))\n",
    "\n",
    "total_aligned_records = total_aligned_records[1:]\n",
    "\n",
    "with open('./data/CLC1-total-aligned.fasta', 'w') as output_handle:\n",
    "    SeqIO.write(total_aligned_records, output_handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b672cf27-f690-4202-a4e1-c1b2558a96ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split into 12 files, each one representing a clade\n",
    "from collections import defaultdict\n",
    "\n",
    "clade_groups = defaultdict(list)\n",
    "\n",
    "with open('./data/CLC1-total-aligned.fasta') as handle:\n",
    "    for seq_record in SeqIO.parse(handle, 'fasta'):\n",
    "        clade_name = seq_record.description.split(' ', 1)[1].split('|')[0].strip()\n",
    "        clade_groups[clade_name].append(seq_record)\n",
    "\n",
    "for clade, seqs in clade_groups.items():\n",
    "    clade = clade.replace('/', '_').replace(' ', '-')\n",
    "    with open(f'./data/CLC1-fastas-by-clade/{clade}_aligned_CLCs.fasta', 'w') as out_f:\n",
    "        SeqIO.write(seqs, out_f, 'fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
